scalgrafanalib/metrics.py

```diff
@@ -0,0 +1,60 @@
+import re
+import typing
+
+class Metric:
+    """Base class for prometheus metric, allowing to statically validate the labels"""
+
+    def __init__(self, name: str, *labels: str, **kwargs: str) -> None:
+        self.name = name
+        self.labels = [*labels, *kwargs.keys()]
+        self.default_labels = self.parse_matchers(**kwargs)
+        self.description: typing.Optional[str] = None
+        self.unit: typing.Optional[str] = None
+
+    def is_valid_label(self, label: str) -> bool:
+        return not self.labels or label in self.labels
+
+    def is_valid_matcher(self, matcher: str) -> bool:
+        match = re.fullmatch('^\\s*([A-Za-z_-]+)\\s*(=|!=|=~|!~)\\s*".*"\\s*$', matcher)
+        return match is not None and self.is_valid_label(match.group(1))
+
+    def parse_matchers(self, *args: str, **kwargs: str) -> typing.List[str]:
+        for arg in args:
+            assert self.is_valid_matcher(
+                arg
+            ), f"Invalid matcher `{arg}` for {self.name}"
+        for kwarg in kwargs:
+            assert self.is_valid_label(
+                kwarg
+            ), f"Invalid label `{kwarg}` for {self.name}"
+        return [*args, *[k + '="' + str(v) + '"' for k, v in kwargs.items()]]
+
+    def with_defaults(self, *args: str, **kwargs: str) -> "Metric":
+        self.default_labels += self.parse_matchers(*args, **kwargs)
+        return self
+
+    def __call__(self, *args: str, **kwargs: str) -> str:
```

Comment:
it is actually the "right" one (obviously opinions may vary...) in this case, as the goal is to use it as a DSL to build prometheus queries, so:
- We can _statically_ check the syntax and label
- We can have default values
- This is written as python, so easier to read/format/ than inside a string
- ...and it is faster, easier and safer to write dashboards thanks to the above 😀

e.g.
```
# Define metrics first (ideally/eventually, this may get auto-generated)
RPO = metrics.BucketMetric(
        'replication_rpo_seconds',
        'location', namespace="${namespace}", job="${data_processor}",
)

# Then build grafana panel
stat = Stat(
   title = "foo",
   targets = Target([
      expr = RPO(location="stuff")
   ])
```

Actual use-case is here: [scality/backbeat@`2267814` (#2323)](https://github.com/scality/backbeat/pull/2323/commits/226781406c655210e72bcb05a0f844c512f337ae)